### ISSUE TITLE
twisted.internet.test.test_tcp.TCPTransportServerAddressTestMixin._testServerAddress hangs for the test method timeout when it fails

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5899]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5899|
|Type|defect|
|Created|2012-08-22 13:24:45Z|
|Branch|https://github.com/twisted/twisted/tree/faster-serveraddresstcp4-failure-5899|

This test method helper relies on `getConnectedClientAndServer` to stop the reactor after the client/server connection is lost.

The connection is closed at the end of the nested `connected` callback inside `_testServerAddress`.  However, if any of the assertions in `connected` fail, the connection won't be closed and the test will hang until the timeout triggers.

<details><summary>Searchable metadata</summary>

```
trac-id__5899 5899
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_faster_serveraddresstcp4_failure_5899 branches/faster-serveraddresstcp4-failure-5899
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1345641885000000 1345641885000000
changetime__1347724059000000 1347724059000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35494]) Branching to 'faster-serveraddresstcp4-failure-5899'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [35495]) Close the connection with a try/finally

refs [#5899](https://github.com/twisted/twisted/issues/5899)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

When this test fails like this:

```

2012-09-03 09:06:45-0400 [Uninitialized] Unhandled Error
        Traceback (most recent call last):
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 551, in _runCallbacks
            current.result = callback(current.result, *args, **kw)
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/test/test_tcp.py", line 1414, in start
            deferred.callback((client, server, port))
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 368, in callback
            self._startRunCallbacks(result)
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 464, in _startRunCallbacks
            self._runCallbacks()
        --- <exception caught here> ---
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/defer.py", line 551, in _runCallbacks
            current.result = callback(current.result, *args, **kw)
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/test/test_tcp.py", line 1335, in connected
            server.transport.logstr)
          File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/unittest.py", line 271, in assertEqual
            % (msg, pformat(first), pformat(second)))
        twisted.trial.unittest.FailTest: not equal:
        a = 'AccumulatingProtocol,0,127.0.0.1'
        b = 'AccumulatingProtocol,0,10.0.0.1'
```

it now does so immediately rather than only after two minutes.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/faster-serveraddresstcp4-failure-5899)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Looks goods, sorry for writing the bad test in the first place... Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [35704]) Merge faster-serveraddresstcp4-failure-5899

Author: exarkun
Reviewer: therve
Fixes: [#5899](https://github.com/twisted/twisted/issues/5899)

Speed up the failure of some `twisted.internet.test.test_tcp` tests, should they fail at all.
